### DISCUSSION TITLE
:hybrid deserialisation flagged as unsafe by brakeman

### DISCUSSION THREAD

**Discussion by ArisBee** [presidentbeef/brakeman#1710] [Q&A] [answered]
Hi there,

I've been scratching my head recently about the [following possible remote code execution](https://brakemanscanner.org/docs/warning_types/unsafe_deserialization/).

From my understanding, while setting the cookie deserialisation to `:json` may be a best practice, I'm surprised the `:hybrid` [type](https://www.bigbinary.com/blog/migrating-existing-session-cookies-while-upgrading-to-rails-4-1-and-above) is also flagged. 
Is there any way this deserialisation could be exploited by a malicious user that would justify to have it flagged?

**Comment by presidentbeef** [comment, marked as answer]
@ArisBee Yes, that's the reason Rails moved from Marshal to JSON for session cookies. `:hybrid` just means "if the session cookie is JSON, accept it. If it's Marshalled, accept it too and convert to JSON."

Crafting a malicious session cookie _does_ require the attacker to know the secret key, though, in order to sign and (depending on Rails version) encrypt the session cookie. Accessing the secret key typically requires another vulnerability... or in the case of open source sometimes people don't change the default secret key grabbed from GitHub.

In any case, it is best not to accept any user-provided input in the Marshal format.
